src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -182,48 +185,65 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
                 terminated = true;
                 break;
             case TokenTypes.LITERAL_BREAK:
-                terminated = useBreak;
+                terminated = useBreak || hasOuterLabel(ast, labels);
                 break;
             case TokenTypes.LITERAL_CONTINUE:
-                terminated = useContinue;
+                terminated = useContinue || hasOuterLabel(ast, labels);
                 break;
             case TokenTypes.SLIST:
-                terminated = checkSlist(ast, useBreak, useContinue);
+                terminated = checkSlist(ast, useBreak, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_IF:
-                terminated = checkIf(ast, useBreak, useContinue);
+                terminated = checkIf(ast, useBreak, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_FOR:
             case TokenTypes.LITERAL_WHILE:
             case TokenTypes.LITERAL_DO:
-                terminated = checkLoop(ast);
+                terminated = checkLoop(ast, labels);
                 break;
             case TokenTypes.LITERAL_TRY:
-                terminated = checkTry(ast, useBreak, useContinue);
+                terminated = checkTry(ast, useBreak, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_SWITCH:
-                terminated = checkSwitch(ast, useContinue);
+                terminated = checkSwitch(ast, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_SYNCHRONIZED:
-                terminated = checkSynchronized(ast, useBreak, useContinue);
+                terminated = checkSynchronized(ast, useBreak, useContinue, labels);
+                break;
+            case TokenTypes.LABELED_STAT:
+                labels.add(ast.getFirstChild().getText());
+                terminated = isTerminated(ast.getLastChild(), useBreak, useContinue, labels);
                 break;
             default:
                 terminated = false;
         }
         return terminated;
     }
 
+    /**
+     * Checks if given break or continue ast has outer label.
+     *
+     * @param ast break or continue node
+     * @param labels label names in scope of case group
+     * @return true if local label used
+     */
+    private static boolean hasOuterLabel(DetailAST ast, Set<String> labels) {
+        final boolean childCount = ast.getChildCount() == 2;
```

Comment:
> Why do we have this condition?

The name could be deceptive. But, this method helps us to avoid fall-through violation. 

```
        top: switch (a) {
            case 1:
                switch (b) {
                    case 10:
                        break top; // ok as terminates outer 'top' switch
                    default:
                        return 11;
                }
            case 2:
                return 2;
        }
```

In the above scenario,` hasOuterLabel` helps us to come out of the entire scope of `top`. And, this prevents fall-through as we don't fall onto case 2 after `break top; `

```
        top: switch (a) {
            case 1:
                inner: switch (b) {
                    case 10:
                        break inner;                     
                        return 11;
                }
            case 2: // Fall through
                return 2;
        }
```

Now, here, we have again a labeled break. We break out of scope of inner switch block and control comes to case 2 causing Fall throgh violation. In nutshell, this method only helps in detecting a labeled break, be it inner or outer.

Reply:
fwiw, I doubt that your second example would compile; we will never reach the `return 11;` statement.


> The name could be deceptive.

Yes, it is. This is not good. 

>childCount

This sounds like it should be a numerical value.


Please do something like this to make it very clear what is happening here:
```suggestion

    private static boolean hasOuterLabel(DetailAST statement, Set<String> labels) {
        final boolean isLabeledStatement = ast.getChildCount() == 2;
        final String identifier = ast.getFirstChild().getText
        return childCount && !labels.contains(identifier);
    }
````

Or to be even more clear and avoid possibility of  `;` being called an identifier or a possible NPE, we could do something like this:
```suggestion

    private static boolean hasOuterLabel(DetailAST statement, Set<String> labels) {       
        return Optional.of(statement)
            .map(DetailAST::getFirstChild)  // get first child
            .filter(child -> child.getType() == TokenTypes.IDENT) // see if it is an identifier
            .map(DetailAST::getText) // if so, get the name
            .filter(label -> !labels.contains(label)) // see if this label is in labels
            .isPresent();
    }

```

Reply:
> fwiw,

Yes, it overlooked it. but, the idea was to explain what haOuterLabel does. it is not a part of the PR.

Reply:
> fwiw, I doubt that your second example would compile; we will never reach the `return 11;` statement.
> 
> > The name could be deceptive.
> 
> Yes, it is. This is not good.
> 
> > childCount
> 
> This sounds like it should be a numerical value.
> 
> Please do something like this to make it very clear what is happening here:
> 
> Or to be even more clear and avoid possibility of `;` being called an identifier or a possible NPE, we could do something like this:

Done.